### PR TITLE
Adding fallback when top level test call is a standardGeneric

### LINKED ISSUE #66: Error when trying to build a test description from a call to an S4 generic
Discovered during the testing of the `Matrix` package, we were unable to generate a test description for a call to a generic that is unavailable to the package namespace:

```r
# Matrix/tests/Class+Meth.R L11:13
setClass("myDGC", contains = "dgCMatrix")
M <- new("myDGC", as(Matrix(c(-2:4, rep(0,9)), 4), "CsparseMatrix"))
M
```

In this case, trying to deparse the top level call throws an error. 

To address this, I think we need a special case for when the call is to a generic.

### PR DESCRIPTION
**Pull request by dgkf**
Closes #66

Will describe details of the patch in review notes below...

NOTE: I haven't been able to install `Matrix` with `--with-keep.source` needed to test the full end-to-end process on my system. I seem to always hit compiler issues when building in a container. This PR addresses any bugs hit before we need the source itself. If anything else pops up it's probably a separate issue.

### CODE REVIEW

File: R/test_description.R [hunk 1]
```diff
@@ -187,6 +187,20 @@ srcref_str <- function(ref) {
...
+      generic, 
+      "(", paste0("<", arg_types, ">", collapse = ", "), ")"
+    )
+
+    return(str)
+  }
+
   if (is.expression(ref)) ref <- as.call(ref)
   gsub("\\s{2,}", " ", paste0(deparse(ref), collapse = "; "))
 }
```
Comment (dgkf) [inline R/test_description.R:206]: This is the crux of the bug fix. Previously, trying to deparse a "standardGeneric" object would throw an error.

Now we first check whether the call is to a standard generic and write out a description of the call and object types that invoked the method.

File: inst/examplepkg/tests/non-testthat.R [hunk 2]
```diff
@@ -0,0 +1,7 @@
+library(examplepkg)
+
+# test scenario where test traces need to be built based on S4 methods 
+# associated with objects defined within the test
+setClass("myS4Example", contains = "S4Example")
+mys4ex <- new("myS4Example", data = list(a = 1, b = 2))  
+mys4ex
```
Comment (dgkf) [inline inst/examplepkg/tests/non-testthat.R:7]: This little snippet reproduces the conditions found in `Matrix` and is used later for testing.

File: tests/testthat/test_test_description.R [hunk 3]
```diff
@@ -104,3 +104,18 @@ test_that("test_description can extract testthat describe-it style test descript
...
+      function(i) class(i[[1]][[1]]) == "standardGeneric",
+      attr(examplepkg_cov, "tests")
+    )
+
+    s4generic_test_desc <- test_description(s4generic_test)
+  })
+
+  expect_equal(attr(s4generic_test_desc, "type"), "call")
+  expect_equal(as.character(s4generic_test_desc), "S4 Generic Call: show(<myS4Example>)")
+})
```
Comment (dgkf) [inline tests/testthat/test_test_description.R:121]: Here I test that we can reproduce a test trace that is a call to a standard generic, and confirm that it produces a reasonable test description.

File: R/pseudo_srcref.R [hunk 4]
```diff
@@ -26,5 +26,15 @@ as.double.with_pseudo_srcref <- function(x, ...) {
 
 #' @export
 format.with_pseudo_srcref <- function(x, ...) {
+  if (inherits(x[[1]], "standardGeneric")) {
+    return(format_standardGeneric_call(x))
+  }
+
   format(unclass(x), ...)
 }
```
Comment (dgkf) [inline R/pseudo_srcref.R:34]: This change isn't strictly necessary to resolve this bug, but handles a downstream issue of not being able to render test srcrefs if they include one of these generic calls. A necessary change to allow our vignettes to continue building.

**Review by maksymiuks — APPROVED** [review]
Great catch and an even better fix! Looking at changes it had to be a nightmare to debug. I hope it didn't cost you too much nerve. Looks really good!

**Pull request merged**